clang/lib/AST/PreorderAST.cpp

```diff
@@ -29,51 +29,48 @@ void PreorderAST::AddNode(Node *N, Node *Parent) {
   }
 }
 
-void PreorderAST::RemoveNode(Node *N, Node *Parent) {
-  // The parent should be a BinaryNode.
-  assert(isa<BinaryNode>(Parent) && "Invalid parent");
+bool PreorderAST::CanCoalesceNode(BinaryNode *B, BinaryNode *P) {
+  if (!B || !isa<BinaryNode>(B) ||
+      !P || !isa<BinaryNode>(P))
+    return false;
+
+  // We can coalesce a BinaryNode if any one of the following conditions are
+  // true:
+  // 1. The parent has the same operator as the current node OR
+  // 2. The current node has just one child (for example, as a result of
+  // constant folding) and the parent and current operators are commutative and
+  // associative.
```

Comment:
My thoughts are as follows:
a) Should the condition 1 above not be "parent has the same operator as the current node, and the operator is commutative and associative" ?
b) If we implement constant folding to be complete in itself, then the current node can never have just one child because of constant folding. It is possible that the current node will have just one child if the operator at the current node is a unary operator.

Reply:
> My thoughts are as follows:
> a) Should the condition 1 above not be "parent has the same operator as the current node, and the operator is commutative and associative" ?
Could you elaborate why this should not be the case? We should not try to coalesce different operators and coalescing only makes sense if the operator is commutative and associative because after coalescing we also need to sort the nodes.

> b) If we implement constant folding to be complete in itself, then the current node can never have just one child because of constant folding. It is possible that the current node will have just one child if the operator at the current node is a unary operator.
We have now implemented constant folding to be complete in itself. As part of constant folding we now invoke this method to coalesce the node. So we need this condition here. Otherwise we would end up duplicating logic to coalesce nodes.